### ISSUE TITLE
Orca Pytorch Estimators should support evaluating validation set during fit

### ISSUE THREAD

**Issue by yushan111** [intel-analytics/BigDL#4584]
Relate to issue #4561 
We need to support evaluating the validation set at each epoch end during training. 

#### With pytorch-lightning
Users could easily evaluate the validation set at the end of each training epoch using `fit`, with passing in `val_dataloader`.
```python
from pytorch_lightning import Trainer

trainer = Trainer()
trainer.fit(model, train_dataloaders, val_dataloaders)
```

#### With Orca TF2 Estimator
```python
from bigdl.orca.learn.tf2 import Estimator

est = Estimator.from_keras(model_creator)
est.fit(data, validation_data)
```

#### With Orca Pytorch Estimator with `bigdl` backend,
```python
from bigdl.orca.learn.pytorch import Estimator

est = Estimator.from_torch(model, backend="bigdl")
est.fit(data, validation_data)
```

### [To be added] Interface for Orca Pytorch Estimator with `ray` and `pyspark` backend
The interface will align with Orca other Estimators, with the seamlessly scale from single-node (e.g. `pytorch-lightning`) to cluster maintained.
```python
from bigdl.orca.learn.pytorch import Estimator

est = Estimator.from_torch(model, backend="ray"/"spark")
est.fit(data, validation_data)
```
Note that users could choose to pass `validation_data` (defaults to `None`) or not. If a valid `validation_data` value is detected, we will automatedly do evaluation after each training epoch, as both `pytorch-lightning` as well as other orca estimators do.

**Comment by jason-dai** [comment]
While this is a nice feature to support, I'm not sure it solves the problem for https://github.com/intel-analytics/BigDL/issues/4561, where the validation results needs to be logged using Weights and Biases:

```python
  def on_train_begin(self):
    self.run = wandb.init(project = self.project, group = "distributed-training")
    pass

  def on_epoch_end(self, epoch):
    out_metrics = self.trainer.validate(self.data_creator)
    self.run.log({"epoch": epoch, "metrics": out_metrics })
```

**Comment by yushan111** [comment]
There might be two requests in issue #4561.
1. Validate at end of each training epoch 
2. Using Weights and Biaes for logging at end of each training epoch.

This issue mainly focuses on the first one. It should be a common use case and I guess we should provide users with a convenient API, since there might be too much labor for users to write a customized callback to achieve that, and it's also unsafe to use our internal API in a customized callback.

For the second one, users could customize a callback. (We may also add a built-in Weights and Biases Callback if many users request the feature). We will also enhance our callbacks interface as described in issue #4586.

So with the new interfaces, to achieve both the two targets, the code could be like
```python
from bigdl.orca.learn.pytorch.callbacks import Callback

class WandbCallback(Callback):
    import wandb
    def __init__(self, project):
        self.run = None
        self.project = project
        super().__init__()
        
    def on_train_begin(self):
        self.run = wandb.init(project = self.project, group = "distributed-training")
    
    def on_epoch_end(self, epoch, logs={}):
        # may log the model weights/gradients retrieved via self.model
        model = self.model
        
        # log the training loss/validation result via logs
        self.run.log({"epoch": epoch})
        self.run.log(logs)        

    def on_train_end(self):
        self.run.finish()

est = Estimator.from_torch(model, backend="ray"/"spark")     
est.fit(data=train_data_creator,
        validation_data=validation_data_creator,
        callbacks=[WandbCallback()]
```

**Comment by jason-dai** [comment]
> So with the new interfaces, to achieve both the two targets, the code could be like
> 
> ```python
> from bigdl.orca.learn.pytorch.callbacks import Callback
> 
> class WandbCallback(Callback):
>     import wandb
>     def __init__(self, project):
>         self.run = None
>         self.project = project
>         super().__init__()
>         
>     def on_train_begin(self):
>         self.run = wandb.init(project = self.project, group = "distributed-training")
>     
>     def on_epoch_end(self, epoch, logs={}):
>         # may log the model weights/gradients retrieved via self.model
>         model = self.model
>         
>         # log the training loss/validation result via logs
>         self.run.log({"epoch": epoch})
>         self.run.log(logs)        
> 
>     def on_train_end(self):
>         self.run.finish()
> 
> est = Estimator.from_torch(model, backend="ray"/"spark")     
> est.fit(data=train_data_creator,
>         validation_data=validation_data_creator,
>         callbacks=[WandbCallback()]
> ```

How does the user log the metrics of the validation results?

**Comment by yushan111** [comment]
> How does the user log the metrics of the validation results?

After fixing issue #4584, `estimator.fit` will take a `validation_data` as input, and return a dictionary containing both the training and validation metric results. E.g: [{"epoch": 1, "training_loss": 0.5, "val_loss": 0.6, "val_accuracy": 0.7}] 

After fixing issue #4586, users could take `logs`, where we have populated both train and validation metric results to, as input of the `on_epoch_end` method in the Callback class. To extract validation metrics only, user could do some filtering like `metric_name.startswith("val_")`
